p2p/src/types.rs

```diff
@@ -192,17 +243,56 @@ impl PeerAddr {
 	/// defaults to port 3414 on mainnet and 13414 on floonet.
 	pub fn from_ip(addr: IpAddr) -> PeerAddr {
 		let port = if global::is_floonet() { 13414 } else { 3414 };
-		PeerAddr(SocketAddr::new(addr, port))
+		PeerAddr::Socket(SocketAddr::new(addr, port))
 	}
 
 	/// If the ip is loopback then our key is "ip:port" (mainly for local usernet testing).
 	/// Otherwise we only care about the ip (we disallow multiple peers on the same ip address).
 	pub fn as_key(&self) -> String {
-		if self.0.ip().is_loopback() {
-			format!("{}:{}", self.0.ip(), self.0.port())
-		} else {
-			format!("{}", self.0.ip())
+		match self {
+			PeerAddr::Socket(s) => {
+				if s.ip().is_loopback() {
+					format!("{}:{}", s.ip(), s.port())
+				} else {
+					format!("{}", s.ip())
+				}
+			}
+			PeerAddr::I2p(i2p_addr) => {
+				i2p_addr.to_string()
+			}
+		}
+	}
+
+	/// Whether this is an i2p address
+	pub fn is_i2p(&self) -> bool {
+		if let PeerAddr::I2p(_) = self {
+			return true;
+		}
+		false
+	}
+
+	/// Whether this is an classic IP address
+	pub fn is_ip(&self) -> bool {
+		if let PeerAddr::I2p(_) = self {
+			return true;
```

Comment:
shouldn't it be `false`? or just replace the body with `!self.is_i2p()`

Reply:
Good catch! Will fix.